### PR TITLE
[BOLT][DWARF][NFC] Split processUnitDIE into two lambdas

### PR DESCRIPTION
**Pull request by sayhaan**
Split processUnitDIE into two lambdas to separate the processing of DWO CUs and CUs in the main binary.

### CODE REVIEW

File: bolt/lib/Rewrite/DWARFRewriter.cpp [hunk 1]
```diff
@@ -768,17 +773,35 @@ void DWARFRewriter::updateDebugInfo() {
...
+      for (DWARFUnit *CU : CompileUnits) {
+        createRangeLocListAddressWriters(*CU, LocListWritersIndexByCU);
+        std::optional<uint64_t> RangesBase;
+        processSplitCU(CU, &DIEBlder, RangesBase);
+        if (RangesBase)
+          RangesBaseByCU[CU->getOffset()] = *RangesBase;
+      }
+      for (DWARFUnit *CU : CompileUnits) {
+        std::optional<uint64_t> RangesBase;
+        if (RangesBaseByCU.count(CU->getOffset()))
```
Comment (ayermolo) [inline bolt/lib/Rewrite/DWARFRewriter.cpp:790]: User iterator to check and get if exists.

File: bolt/lib/Rewrite/DWARFRewriter.cpp [hunk 2]
```diff
@@ -632,120 +632,125 @@ void DWARFRewriter::updateDebugInfo() {
...
+            RangeListsWritersByCU[*DWOId] =
+                std::move(DWORangeListsSectionWriter);
+          }
+          AddressWritersByCU[CU.getOffset()] = std::move(AddrW);
+
+        } else {
+          auto AddrW =
+              std::make_unique<DebugAddrWriter>(&BC, CU.getAddressByteSize());
+          AddressWritersByCU[CU.getOffset()] = std::move(AddrW);
+          LocListWritersByCU[CUIndex] = std::make_unique<DebugLocWriter>();
```
Comment (ayermolo) [inline bolt/lib/Rewrite/DWARFRewriter.cpp:663]: Maybe it's best to simplify this and just created instance of Loc/loclist writer when we iterate for Main Binary and store it in vector.

File: bolt/lib/Rewrite/DWARFRewriter.cpp [hunk 3]
```diff
@@ -632,120 +632,125 @@ void DWARFRewriter::updateDebugInfo() {
...
+                             Unit);
+    DWODIEBuilder.buildDWOUnit(**SplitCU);
+    std::string DWOName = "";
+    std::optional<std::string> DwarfOutputPath =
+        opts::DwarfOutputPath.empty()
+            ? std::nullopt
+            : std::optional<std::string>(opts::DwarfOutputPath.c_str());
+    {
+      std::lock_guard<std::mutex> Lock(AccessMutex);
+      DWOName = DIEBlder->updateDWONameCompDir(
```
Comment (ayermolo) [inline bolt/lib/Rewrite/DWARFRewriter.cpp:703]: This common state that will require lock. In followup diff move it outside of this lambda.

File: bolt/lib/Rewrite/DWARFRewriter.cpp [hunk 4]
```diff
@@ -632,120 +632,125 @@ void DWARFRewriter::updateDebugInfo() {
...
+                                         AddressWriter);
+    DebugRangesSectionWriter *TempRangesSectionWriter;
+    if (Unit->getVersion() >= 5) {
+      TempRangesSectionWriter = RangeListsWritersByCU[*DWOId].get();
+    } else {
+      TempRangesSectionWriter = LegacyRangesWritersByCU[*DWOId].get();
+      RangesBase = Unit->getVersion() >= 5
+                       ? RangeListsSectionWriter.get()->getSectionOffset()
+                       : LegacyRangesSectionWriter.get()->getSectionOffset();
+      setDwoRangesBase(*DWOId, *RangesBase);
```
Comment (ayermolo) [inline bolt/lib/Rewrite/DWARFRewriter.cpp:720]: Hmm actually this whole setDwoRangesBase etc no longer used anywhere....
Can you remove it in a different PR. Rebase this one on top of it.

File: bolt/lib/Rewrite/DWARFRewriter.cpp [hunk 5]
```diff
@@ -632,120 +632,125 @@ void DWARFRewriter::updateDebugInfo() {
...
+    DWODIEBuilder.updateDWONameCompDirForTypes(
+        DWOStrOffstsWriter, DWOStrWriter, **SplitCU, DwarfOutputPath, DWOName);
+    DebugLoclistWriter DebugLocDWoWriter(*Unit, Unit->getVersion(), true,
+                                         AddressWriter);
+    DebugRangesSectionWriter *TempRangesSectionWriter;
+    if (Unit->getVersion() >= 5) {
+      TempRangesSectionWriter = RangeListsWritersByCU[*DWOId].get();
+    } else {
+      TempRangesSectionWriter = LegacyRangesWritersByCU[*DWOId].get();
+      RangesBase = Unit->getVersion() >= 5
```
Comment (ayermolo) [inline bolt/lib/Rewrite/DWARFRewriter.cpp:717]: I don't think RangesBase needs to be set here. It's only needed when we process CUs in main binary, so we know when to update it, or create a new one.

File: bolt/lib/Rewrite/DWARFRewriter.cpp (hunk 2 above)
Comment (sayhaan) [reply bolt/lib/Rewrite/DWARFRewriter.cpp:663]: I'm not sure if we can do that because the instance is created in createRangeLocListAddressWriters which also populates the Address/Ranges writers maps and uses the Address Writer, which processSplitCU uses. If move it to the loop where we iterate for the main binary then processSplitCU won't work.

Comment (ayermolo) [reply bolt/lib/Rewrite/DWARFRewriter.cpp:663]: Ah right, dwarf5 instance takes in.

File: bolt/lib/Rewrite/DWARFRewriter.cpp [hunk 6]
```diff
@@ -767,17 +769,28 @@ void DWARFRewriter::updateDebugInfo() {
...
       finalizeCompileUnits(DIEBlder, *Streamer, OffsetMap,
                            DIEBlder.getProcessedCUs(), *FinalAddrWriter);
     }
   } else {
     // Update unit debug info in parallel
     ThreadPoolInterface &ThreadPool = ParallelUtilities::getThreadPool();
     for (std::unique_ptr<DWARFUnit> &CU : BC.DwCtx->compile_units())
-      ThreadPool.async(processUnitDIE, CU.get(), &DIEBlder);
-    ThreadPool.wait();
+      // ThreadPool.async(processUnitDIE, CU.get(), &DIEBlder);
```
Comment (ayermolo) [inline bolt/lib/Rewrite/DWARFRewriter.cpp:794]: Remove this change

File: bolt/lib/Rewrite/DWARFRewriter.cpp [hunk 7]
```diff
@@ -767,17 +769,28 @@ void DWARFRewriter::updateDebugInfo() {
     CUPartitionVector PartVec = partitionCUs(*BC.DwCtx);
     for (std::vector<DWARFUnit *> &Vec : PartVec) {
       DIEBlder.buildCompileUnits(Vec);
-      for (DWARFUnit *CU : DIEBlder.getProcessedCUs())
-        processUnitDIE(CU, &DIEBlder);
+      llvm::SmallVector<DWARFUnit *> CompileUnits(
```
Comment (ayermolo) [inline bolt/lib/Rewrite/DWARFRewriter.cpp:774]: Why is this necessary?

File: bolt/lib/Rewrite/DWARFRewriter.cpp [hunk 8]
```diff
@@ -632,119 +632,121 @@ void DWARFRewriter::updateDebugInfo() {
...
+    DebugLocDWoWriter.finalize(DWODIEBuilder,
+                               *DWODIEBuilder.getUnitDIEbyUnit(**SplitCU));
+    if (Unit->getVersion() >= 5)
+      TempRangesSectionWriter->finalizeSection();
 
+    emitDWOBuilder(DWOName, DWODIEBuilder, *this, **SplitCU, *Unit, State,
+                   DebugLocDWoWriter, DWOStrOffstsWriter, DWOStrWriter,
+                   GDBIndexSection);
+  };
+  auto processMainBinaryCU = [&](DWARFUnit *Unit, DIEBuilder *DIEBlder,
```
Comment (ayermolo) [inline bolt/lib/Rewrite/DWARFRewriter.cpp:726]: Since we are modifying this code anyway. In this and other lambda change to pass by reference.

File: bolt/lib/Rewrite/DWARFRewriter.cpp [hunk 9]
```diff
@@ -632,119 +632,121 @@ void DWARFRewriter::updateDebugInfo() {
...
+                   "RangeLists writer for DWO unit already exists.");
+            auto DWORangeListsSectionWriter =
+                std::make_unique<DebugRangeListsSectionWriter>();
+            DWORangeListsSectionWriter->initSection(CU);
+            DWORangeListsSectionWriter->setAddressWriter(AddrW.get());
+            RangeListsWritersByCU[*DWOId] =
+                std::move(DWORangeListsSectionWriter);
+          }
+          AddressWritersByCU[CU.getOffset()] = std::move(AddrW);
+
```
Comment (ayermolo) [inline bolt/lib/Rewrite/DWARFRewriter.cpp:658]: remove space

File: bolt/lib/Rewrite/DWARFRewriter.cpp (hunk 7 above)
Comment (sayhaan) [reply bolt/lib/Rewrite/DWARFRewriter.cpp:774]: Since we have two loops we only call DIEBlder.getProcessedCUs() once instead of in both loops.

File: bolt/lib/Rewrite/DWARFRewriter.cpp [hunk 10]
```diff
@@ -621,119 +621,122 @@ void DWARFRewriter::updateDebugInfo() {
...
+    DIEBuilder DWODIEBuilder(BC, &(*SplitCU)->getContext(), DebugNamesTable,
+                             &Unit);
+    DWODIEBuilder.buildDWOUnit(**SplitCU);
+    std::string DWOName = "";
+    std::optional<std::string> DwarfOutputPath =
+        opts::DwarfOutputPath.empty()
+            ? std::nullopt
+            : std::optional<std::string>(opts::DwarfOutputPath.c_str());
+    {
+      std::lock_guard<std::mutex> Lock(AccessMutex);
```
Comment (ayermolo) [inline bolt/lib/Rewrite/DWARFRewriter.cpp:689]: Remove in followup PR.

File: bolt/lib/Rewrite/DWARFRewriter.cpp [hunk 11]
```diff
@@ -621,119 +621,122 @@ void DWARFRewriter::updateDebugInfo() {
...
-                                 *DWODIEBuilder.getUnitDIEbyUnit(**SplitCU));
-      if (Unit->getVersion() >= 5)
-        TempRangesSectionWriter->finalizeSection();
-
-      emitDWOBuilder(DWOName, DWODIEBuilder, *this, **SplitCU, *Unit, State,
-                     DebugLocDWoWriter, DWOStrOffstsWriter, DWOStrWriter,
-                     GDBIndexSection);
+    if (!SplitCU)
+      return;
+    DebugAddrWriter &AddressWriter =
```
Comment (ayermolo) [inline bolt/lib/Rewrite/DWARFRewriter.cpp:678]: Move the checks outside, and just pass in AddressWriter and SplitCU.
My thinking is less reliance on shared data structures the better.

File: bolt/lib/Rewrite/DWARFRewriter.cpp [hunk 12]
```diff
@@ -621,119 +621,122 @@ void DWARFRewriter::updateDebugInfo() {
...
-
-    if (Unit->getVersion() >= 5) {
-      RangesBase = RangesSectionWriter->getSectionOffset() +
+    DebugStrOffsetsWriter DWOStrOffstsWriter(BC);
+    DebugStrWriter DWOStrWriter((*SplitCU)->getContext(), true);
+    DWODIEBuilder.updateDWONameCompDirForTypes(
+        DWOStrOffstsWriter, DWOStrWriter, **SplitCU, DwarfOutputPath, DWOName);
+    DebugLoclistWriter DebugLocDWoWriter(Unit, Unit.getVersion(), true,
+                                         AddressWriter);
+    DebugRangesSectionWriter *TempRangesSectionWriter =
```
Comment (ayermolo) [inline bolt/lib/Rewrite/DWARFRewriter.cpp:699]: Also this, move check outside of lambda and pass in.

File: bolt/lib/Rewrite/DWARFRewriter.cpp [hunk 13]
```diff
@@ -621,119 +621,122 @@ void DWARFRewriter::updateDebugInfo() {
...
-               "RangeLists writer for DWO unit already exists.");
-        auto DWORangeListsSectionWriter =
-            std::make_unique<DebugRangeListsSectionWriter>();
-        DWORangeListsSectionWriter->initSection(CU);
-        DWORangeListsSectionWriter->setAddressWriter(AddrW.get());
-        RangeListsWritersByCU[*DWOId] = std::move(DWORangeListsSectionWriter);
-      }
-      AddressWritersByCU[CU.getOffset()] = std::move(AddrW);
+      [&](DWARFUnit &CU,
+          llvm::DenseMap<uint64_t, uint64_t> &LocListWritersIndexByCU) {
```
Comment (ayermolo) [inline bolt/lib/Rewrite/DWARFRewriter.cpp:625]: I think you can just capture it by reference in lambda.

Comment (sayhaan) [reply bolt/lib/Rewrite/DWARFRewriter.cpp:625]: For LocListWritersIndexByCU? Wouldn't it not be in the same scope since I initialize it inside the for loop? I could move it outside but that would increase the scope it covers when it's only used in one lambda.

Comment (sayhaan) [reply bolt/lib/Rewrite/DWARFRewriter.cpp:625]: For LocListWritersIndexByCU? It wouldn't be in scope for the lambda to capture by reference since I initialize it inside the for loop. I could move it outside but then its lifetime is longer than it needs to be since it's only used in the lambda.

**Review by ayermolo — APPROVED** [review]

**Comment by aaupov** [comment]
Heads up that this diff is not NFC: https://lab.llvm.org/buildbot/#/builders/92/builds/2042
